File: examples/canonical-auth/go/internal/domain/store.go
package domain

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

type InMemoryStore struct {
	Users           []User
	Tenants         []Tenant
	Memberships     []Membership
	Groups          []Group
	Permissions     []Permission
	GroupPermissions []GroupPermission
	UserGroups      []UserGroup
}

func (s *InMemoryStore) GetUserByID(id string) (*User, bool) {
	for _, u := range s.Users {
		if u.ID == id {
			return &u, true
		}
	}
	return nil, false
}

func (s *InMemoryStore) GetUserByEmail(email string) (*User, bool) {
	for _, u := range s.Users {
		if u.Email == email {
			return &u, true
		}
	}
	return nil, false
}

func (s *InMemoryStore) GetTenantByID(id string) (*Tenant, bool) {
	for _, t := range s.Tenants {
		if t.ID == id {
			return &t, true
		}
	}
	return nil, false
}

func (s *InMemoryStore) GetMembership(userID string) (*Membership, bool) {
	for _, m := range s.Memberships {
		if m.UserID == userID && m.IsActive {
			return &m, true
		}
	}
	return nil, false
}

func (s *InMemoryStore) GetGroupsForUser(userID, tenantID string) []Group {
	var userGroupIDs []string
	for _, ug := range s.UserGroups {
		if ug.UserID == userID {
			userGroupIDs = append(userGroupIDs, ug.GroupID)
		}
	}

	var groups []Group
	for _, g := range s.Groups {
		if g.TenantID == tenantID {
			for _, ugID := range userGroupIDs {
				if g.ID == ugID {
					groups = append(groups, g)
					break
				}
			}
		}
	}
	return groups
}

func (s *InMemoryStore) GetEffectivePermissions(userID string) []string {
	var userGroupIDs []string
	for _, ug := range s.UserGroups {
		if ug.UserID == userID {
			userGroupIDs = append(userGroupIDs, ug.GroupID)
		}
	}

	var groupIDs []string
	for _, g := range s.Groups {
		for _, ugID := range userGroupIDs {
			if g.ID == ugID {
				groupIDs = append(groupIDs, g.ID)
				break
			}
		}
	}

	var permIDs []string
	for _, gp := range s.GroupPermissions {
		for _, gID := range groupIDs {
			if gp.GroupID == gID {
				permIDs = append(permIDs, gp.PermissionID)
				break
			}
		}
	}

	permMap := make(map[string]bool)
	for _, p := range s.Permissions {
		for _, pID := range permIDs {
			if p.ID == pID {
				permMap[p.Name] = true
				break
			}
		}
	}

	var perms []string
	for p := range permMap {
		perms = append(perms, p)
	}
	return perms
}

func (s *InMemoryStore) VerifyMembership(userID, tenantID string) bool {
	for _, m := range s.Memberships {
		if m.UserID == userID && m.IsActive {
			if m.TenantID != nil && *m.TenantID == tenantID {
				return true
			}
		}
	}
	return false
}

func (s *InMemoryStore) GetTenantName(tenantID string) string {
	if t, ok := s.GetTenantByID(tenantID); ok {
		return t.Name
	}
	return ""
}

func LoadFromFixtures(fixtureDir string) (*InMemoryStore, error) {
	if fixtureDir == "" {
		fixtureDir = "../../domain/fixtures"
	}

	store := &InMemoryStore{}

	if err := loadJSON(filepath.Join(fixtureDir, "users.json"), &store.Users); err != nil {
		return nil, err
	}
	if err := loadJSON(filepath.Join(fixtureDir, "tenants.json"), &store.Tenants); err != nil {
		return nil, err
	}
	if err := loadJSON(filepath.Join(fixtureDir, "memberships.json"), &store.Memberships); err != nil {
		return nil, err
	}
	if err := loadJSON(filepath.Join(fixtureDir, "groups.json"), &store.Groups); err != nil {
		return nil, err
	}
	if err := loadJSON(filepath.Join(fixtureDir, "permissions.json"), &store.Permissions); err != nil {
		return nil, err
	}
	if err := loadJSON(filepath.Join(fixtureDir, "group_permissions.json"), &store.GroupPermissions); err != nil {
		return nil, err
	}
	if err := loadJSON(filepath.Join(fixtureDir, "user_groups.json"), &store.UserGroups); err != nil {
		return nil, err
	}

	return store, nil
}

func loadJSON(path string, v interface{}) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("failed to read fixture %s: %w", path, err)
	}
	if err := json.Unmarshal(data, v); err != nil {
		return fmt.Errorf("failed to parse fixture %s: %w", path, err)
	}
	return nil
}

File: examples/canonical-auth/rust/src/domain/store.rs
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::Path;

use super::models::*;

#[derive(Debug, Clone)]
pub struct InMemoryStore {
    pub users: Vec<User>,
    pub tenants: Vec<Tenant>,
    pub memberships: Vec<Membership>,
    pub groups: Vec<Group>,
    pub permissions: Vec<Permission>,
    pub group_permissions: Vec<GroupPermission>,
    pub user_groups: Vec<UserGroup>,
}

impl InMemoryStore {
    pub fn load_from_fixtures(fixture_dir: &Path) -> Result<Self, Box<dyn std::error::Error>> {
        let users: Vec<User> = serde_json::from_slice(&fs::read(fixture_dir.join("users.json"))?)?;
        let tenants: Vec<Tenant> = serde_json::from_slice(&fs::read(fixture_dir.join("tenants.json"))?)?;
        let memberships: Vec<Membership> = serde_json::from_slice(&fs::read(fixture_dir.join("memberships.json"))?)?;
        let groups: Vec<Group> = serde_json::from_slice(&fs::read(fixture_dir.join("groups.json"))?)?;
        let permissions: Vec<Permission> = serde_json::from_slice(&fs::read(fixture_dir.join("permissions.json"))?)?;
        let group_permissions: Vec<GroupPermission> = serde_json::from_slice(&fs::read(fixture_dir.join("group_permissions.json"))?)?;
        let user_groups: Vec<UserGroup> = serde_json::from_slice(&fs::read(fixture_dir.join("user_groups.json"))?)?;

        Ok(Self {
            users,
            tenants,
            memberships,
            groups,
            permissions,
            group_permissions,
            user_groups,
        })
    }

    pub fn get_user_by_id(&self, id: &str) -> Option<&User> {
        self.users.iter().find(|u| u.id == id)
    }

    pub fn get_user_by_email(&self, email: &str) -> Option<&User> {
        self.users.iter().find(|u| u.email == email)
    }

    pub fn get_tenant_by_id(&self, id: &str) -> Option<&Tenant> {
        self.tenants.iter().find(|t| t.id == id)
    }

    pub fn get_tenant_name(&self, tenant_id: &str) -> Option<&str> {
        self.get_tenant_by_id(tenant_id).map(|t| t.name.as_str())
    }

    pub fn get_membership(&self, user_id: &str) -> Option<&Membership> {
        self.memberships.iter().find(|m| m.user_id == user_id && m.is_active)
    }

    pub fn verify_membership(&self, user_id: &str, tenant_id: &str) -> bool {
        self.memberships.iter().any(|m| {
            m.user_id == user_id && m.is_active && m.tenant_id.as_deref() == Some(tenant_id)
        })
    }

    pub fn get_groups_for_user(&self, user_id: &str, tenant_id: &str) -> Vec<&Group> {
        let user_group_ids: HashSet<_> = self.user_groups.iter()
            .filter(|ug| ug.user_id == user_id)
            .map(|ug| &ug.group_id)
            .collect();

        self.groups.iter()
            .filter(|g| g.tenant_id == tenant_id && user_group_ids.contains(&g.id))
            .collect()
    }

    pub fn get_effective_permissions(&self, user_id: &str) -> Vec<String> {
        let user_group_ids: HashSet<_> = self.user_groups.iter()
            .filter(|ug| ug.user_id == user_id)
            .map(|ug| &ug.group_id)
            .collect();

        let group_ids: HashSet<_> = self.groups.iter()
            .filter(|g| user_group_ids.contains(&g.id))
            .map(|g| &g.id)
            .collect();

        let perm_ids: HashSet<_> = self.group_permissions.iter()
            .filter(|gp| group_ids.contains(&gp.group_id))
            .map(|gp| &gp.permission_id)
            .collect();

        let mut perms: Vec<String> = self.permissions.iter()
            .filter(|p| perm_ids.contains(&p.id))
            .map(|p| p.name.clone())
            .collect();
        perms.sort();
        perms
    }
}
